### ISSUE TITLE
New vignettes would be greatly appreciated! I had the previous version of stemr working well for generating SEMs for the COVID outbreak based on death incidence, but the update to Ryacas broke my workflow.

### ISSUE THREAD

**Issue by fintzij** [fintzij/stemr#9]
New vignettes would be greatly appreciated! I had the previous version of stemr working well for generating SEMs for the COVID outbreak based on death incidence, but the update to Ryacas broke my workflow.

If new vignettes are not imminent, perhaps I could get the answers to a couple of questions?
1. For the priors input to the function parblocks, is logprior the same function as what you were calling prior_density in previous versions? If not, how is logprior specified?
2. Is the intializer input to parblocks used to initialize the parameters in the same manner as the function assigned to stemobj$dynamics$parameters in previous versions? (i.e. to add a random perturbation to the starting parameter value)

These were my assumptions after reading the commented code, but I get an initialization error when fit_stem tries to call initialize_lna.

_Originally posted by @kevinkiernan in https://github.com/fintzij/stemr/issues/3#issuecomment-691399795_

**Comment by fintzij** [comment]
Hi Kevin,

Sorry, I'll try to get a vignette posted this week. If you're in a pinch and need an example sooner, you could grab one from the .development_files folder within the examples_stemr_0.2 subdirectory. The user-facing functions in the package should be stable in the short term, although there are still outstanding issues that might affect how users interact with the package. I'm not planning on more refactoring in the next few months, but apologies in advance. I'll make sure to update vignettes when pushing user-facing updates in the future.

If you're having Ryacas issues, I would try reinstalling stemr and updating Ryacas. The previous 0.1 version depended on an Ryacas function that was deprecated. This meant that stemr was reliant on an old version, which wasn't really ideal. Maybe updating both packages will solve your issue. Also, no need to install without byte compilation any more as that problem is more or less sorted out. 

The answer to both of your questions is yes. logprior is used in the same way that prior_density previously was, except that you should no longer supply params_nat as an argument, just params_est. The initializer argument also works in exactly the same way, it's just supplied in a different place. A few questions: 1) Are you still having initialization issues if you fix the initial parameter values? 2) It's generally better to randomly initialize parameters on the estimation scale and transform back to the natural scale, as in the examples. This is especially true if you have any boundary conditions you want to avoid or if there are certain parameters that would break the numerical ODE solvers, e.g., R0 = 100. 3) Which approximation are you using in fitting your models - ODE, LNA with exact initialization, LNA with approximate initialization + warmup? Does it break for all of the approximations?

Cheers,
Jon

**Comment by kevinkiernan** [comment]
Hi Jon,

Thanks very much for responding. I was able to get back on track by rolling back to stemr version 0.1.0 and Ryacas version 0.4.1

Regarding your questions, I am initializing the parameter values on the estimation scale as you say, and I have attempted both ODE and LNA with warmup. The parameters appear to be well in bounds (I'm having the terminal print them as they're generated), so I believe I'm defining some input to the model incorrectly. I had a similar set of issues when I first tried to run the package--the from_estimation function was generating NaN values.

For now, I will continue working with the 0.1.0 version and await vignettes as you are able to post them. 

--Kevin

**Comment by fintzij** [comment]
Sounds like a plan. I'm still curious as to why it's throwing up errors. I'd be happy to read through and try to find errors if you want to post commented code.

Cheers,
Jon

**Comment by fintzij** [comment]
Hey Kevin, I just pushed a vignette for the SIR model. I'll wait to close this issue until we resolve whatever bug you're running into.

**Comment by kevinkiernan** [comment]
Hey Jon, that's wonderful. I'll review it as soon as I have time and let you know if the issues persist. It might be a few days--I hope that's ok.

**Comment by fintzij** [comment]
No worries, take your time

**Comment by kevinkiernan** [comment]
Hi Jon,

I'm unfortunately still experiencing problems with v0.2.0, and the code in the vignettes looks more or less the same as what's in my scripts. I'm posting my code and the underlying data in the hope you can spot what is going on. 

Stepping back a bit, it's probably worth explaining what my project is about. I'm using the code to reconstruct the early stages of the COVID outbreak from death incidence. It's pretty clear that case counts have significant data quality issues, so I'm attempting to back out the compartments of a SEIRD model using stemr. I've got v0.1.0 running in a separate Docker container, so it's not currently causing me active problems.

Thanks, 
Kevin
[KHK_stemr_files.zip](https://github.com/fintzij/stemr/files/5240972/KHK_stemr_files.zip)

**Comment by fintzij** [comment]
Thanks Kevin. Give me a day or two to dig into this. My guess is that it likely has something to do with the specification for the time-varying reproduction number. A couple of quick notes:
- the definition of beta_t0 on line 66 shouldn't include constants["sigma"] in this model. R0 = beta * N / delta for your SEIRD model.
- You've got more random variates in the n_draws argument to the time-varying parameter block than time intervals. I'm almost positive that that's going to cause problems. n_draws should correspond to the number of random variates that you want to be jointly sampled via elliptical slice sampling. This will generally be equal to the number of intervals + the number of hyperparameters if you're sampling those jointly and they are normally distributed a priori. I'd suggest fixing the standard deviation of the random walk instead of estimating it, just as a start.
- Those spikes in the data are probably case backlogs and are going to wreck the computation. We don't have functionality for reporting delays built in yet, but a poor man's solution might be to censor those big spikes and have 2-3 parameters for the mean death detection rate, once before and once after each spike at weeks 32/33 and 39. You could even estimate those as time-varying parameters and do some smoothing.
- Strongly suggest parameterizing the estimation scale in terms of log(R0_t0) instead of log(beta_t0). You might also consider a more informative prior on the initial reproduction number since quite a lot is known about the scale of R0 at that time, e.g., it most assuredly was not less than 1 or greater than 10. As it is, 90% of your prior mass is between 0.004 and 53000, which is kinda crazy.
- Biologically, the IFR has to be between 0 and 1. So, you should probably logit transform omega to get it on its estimation scale. The prior on the IFR also isn't going to do you any big favors. It'll make sense to tighten it up so that, at the very least, the prior mass is less than the crude case fatality rate and whatever value for IFR would imply that no more people were infected than live in the whole state of New York. Another problem with a super diffuse prior on the logit scale is that the induced prior on the probability scale ends up looking like a point mass at 0 and 1, which is probably not what you want to induce even if you mean for your prior to be agnostic. 

Cheers,
Jon

**Comment by kevinkiernan** [comment]
Jon,

Thanks for the comments--I was aware of several of them and had intended to circle back around to fixing them once I got the code to successfully run. I will say that most of these issues did not cause initialization errors in the previous version.

The n_draws argument appears to be part of the issue I'm having. I've managed to get the code to initialize by setting n_draws = length(times). I will set it to n_draws = length(times) + length(parameters) per your suggestion. Also I had conflicting statements about the control algorithm of the parameter blocks. Something still appears not to be quite right, as indicated by the code bogging down at the "Beginning MCMC" stage. 

I'll prioritize your suggestions and see where that leaves me.

--Kevin

**Comment by kevinkiernan** [comment]
This code implements most of your suggestions, including:
--Parameterization in terms of R_t instead of beta_t (using the corrected expression for R_t)
--Fixed the n_draws argument in the tparam definition
--Logit transforming the IFR

The n_draws argument was the source of the error I was experiencing previously, but now the code runs for a couple of hours and returns empty results. Specifically, the matrix of tparam draws is just zeros. 

I've checked and feeding the function foi_rw1 the parameters and a vector of normally distributed draws produces a random walk like it should.

Both the input script and the output are attached. Anything jump out at you about why this might be happening?

Thanks,
Kevin
[KHK_stemr_files2.zip](https://github.com/fintzij/stemr/files/5248866/KHK_stemr_files2.zip)

**Comment by fintzij** [comment]
I think there was still an error in the specification of the GMRF. You're not actually sampling any parameters along with the time varying reproduction numbers, but you added some extra draws for that. Also, it'll be a bit faster to back out beta_t with time-varying parameters instead of the rates. This way you don't need to back out beta_t every time the ODE solvers take a step. 

Anyway, the attached code seems to work for me. I tried it with the ODE, not the LNA approximation. The functions to save the posterior samples for time-varying parameters and their non-centered draws are independent of the latent process approximation, so i think you should be good to go with the LNA as well.

Cheers,
Jon

[KHK_stemr_files3.zip](https://github.com/fintzij/stemr/files/5250532/KHK_stemr_files3.zip)

**Comment by fintzij** [comment]
Oh, also, there was a minor bug in the wrapper for the MCMC. I was accidentally returning an extra empty sample. If you reinstall the package the bugfix will get rid of that.

Cheers,
Jon

**Comment by kevinkiernan** [comment]
This worked, although with a wrinkle you may find interesting: even after reinstalling stemr, the code produced the same empty results file as I sent you, until I upgraded my R version to 4.0.2 (I had been running 3.6.1). After that it worked well.

Thanks for your help and suggestions!

--Kevin

**Comment by fintzij** [comment]
That is so strange! Glad it's working now.

Cheers,
Jon
